### PR TITLE
Fixes for JENKINS-27070 + JENKINS-16703

### PR DESCRIPTION
**Pull request by aheritier**
For JENKINS-27070 I just upgraded the crowd library to have a newer version that uses HTTPComponents 4.x
For JENKINS-16703 I centralized all crowd calls in CrowdConfigurationServices and added a cache mechanism like in the LDAP plugin to cache the data retrieved from crowd. The administrator can globally configure the number of entries and the TTL of all caches.

### CODE REVIEW

File: src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.jelly [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+The MIT License
+
+Copyright 2014 Jesse Glick.
```
Comment (KostyaSha) [inline src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.jelly]: @jglick really?

**Comment by KostyaSha** [comment]
Please read how contribute to jenkins, code full of coding style changes, i can't find any real changes.
{code}
jenkinsadmin commented 14 hours ago
Thank you for a pull request! Please check this document for how the Jenkins project handles pull requests
{code}

**Comment by KostyaSha** [comment]
Have you tested that 1.509.3 jenkins works after this changes?

File: src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.jelly (hunk 1 above)
Comment (aheritier) [inline src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.jelly]: LOL. A too quick copy/paste. I'll fix it.

Le samedi 4 avril 2015, Kanstantsin Shautsou <notifications@github.com> a
écrit :

> In
> src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.jelly
> <https://github.com/jenkinsci/crowd2-plugin/pull/10#discussion_r27767473>:
>
> > @@ -0,0 +1,34 @@
> > +<?xml version="1.0" encoding="UTF-8"?>
> > +<!--
> > +The MIT License
> > +
> > +Copyright 2014 Jesse Glick.
>
> @jglick <https://github.com/jglick> really?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/crowd2-plugin/pull/10/files#r27767473>.
>


-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier

File: src/main/java/de/theit/jenkins/crowd/CrowdConfigurationService.java [hunk 2]
```diff
@@ -25,335 +25,658 @@
...
+     *
+     * @param username The name of the user to check. May not be <code>null</code> or
+     *                 empty.
+     * @return <code>true</code> if and only if the group exists, is active and
+     * the user is either a direct group member or, if nested groups may
+     * be used, a nested group member. <code>false</code> else.
+     */
+    public boolean isGroupMember(String username) {
+        if (allowedGroupNames.isEmpty()) {
+            return true;
```
Comment (KostyaSha) [inline src/main/java/de/theit/jenkins/crowd/CrowdConfigurationService.java]: You removed
```
-		if (! this.configuration.allowedGroupNames.isEmpty()) {
``

**Comment by aheritier** [comment]
That's mainly impossible with some shared coding style. I asked on irc to
see if there was some rules nowadays and the reply is no. That's a pity.
We clearly cannot have one formatting rule per plugin.
 I often tried to reformat only the code I worked on but it's not easy.
If you have some rules for eclipse or IntelliJ please share them, I'll
happily reformat the code to have a readable diff

Le samedi 4 avril 2015, Kanstantsin Shautsou <notifications@github.com> a
écrit :

> Please read how contribute to jenkins, code full of coding style changes,
> i can't find any real changes.
> {code}
> jenkinsadmin commented 14 hours ago
> Thank you for a pull request! Please check this document for how the
> Jenkins project handles pull requests
> {code}
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/crowd2-plugin/pull/10#issuecomment-89518012>
> .
>


-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier

**Comment by KostyaSha** [comment]
There is only one rule - don't modify code that is not related to your change.

**Comment by aheritier** [comment]
In Dev mode locally yes. It works. In production only with the latest LTS.
There is only the SSO I don't succeed to use but it was always the case. I
probably have a wrong setup somewhere in a reverse proxy. I deactivated it
myself because it generates too much useless requests.

Le samedi 4 avril 2015, Kanstantsin Shautsou <notifications@github.com> a
écrit :

> Have you tested that 1.509.3 jenkins works after this changes?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/crowd2-plugin/pull/10#issuecomment-89520782>
> .
>


-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier

**Comment by KostyaSha** [comment]
What is dev mode? hpi:run + connect it to some crowd.

**Comment by aheritier** [comment]
Yes exactly.

Le samedi 4 avril 2015, Kanstantsin Shautsou <notifications@github.com> a
écrit :

> What is dev mode? hpi:run + connect it to some crowd.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/crowd2-plugin/pull/10#issuecomment-89523790>
> .
>


-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier

**Comment by KostyaSha** [comment]
I will need test some cases (i expect class conflicts with other plugins), review changes in library and more time if coding style change will be still here. Also massive coding style change will broke my refactoring work...

**Comment by aheritier** [comment]
Don't be worry. I just wanted to have a quick fix and I shared it as soon
as possible. I only spent one day on it.
I won't be able to work on it in next 10 days but I'll propose a cleaner PR
when I'll be back from Devoxx France. It will give us also more feedback
from the 2 production servers where I deployed my patched version.

Cheers.

Le samedi 4 avril 2015, Kanstantsin Shautsou <notifications@github.com> a
écrit :

> I will need test some cases (i expect class conflicts with other plugins),
> review changes in library and more time if coding style change will be
> still here. Also massive coding style change will broke my refactoring
> work...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/crowd2-plugin/pull/10#issuecomment-89528737>
> .
>


-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier

**Comment by KostyaSha** [comment]
Check isGroupMember function because it looks like security flow.

**Comment by aheritier** [comment]
Hi @KostyaSha,

  The good news is that my patched version deployed 10 days ago is working fine from the end-users point of view. My coworkers never seen our Jenkins so reactive. I didn't have any production issue (memory consumption problem ...).
  The bad news is that I didn't find the time this week to redo the patch to propose you a sanitised version. Next week I'll be off thus it won't land before 2 weeks min. (Note for sure I'll have more time in May to work on Jenkins - I will join the Cloudbees support team).
  About the PR do you prefer it on top of master or on top of your refactoring branch ?

Cheers,

**Comment by KostyaSha** [comment]
@aheritier if you are really interested in having working rememberme, then i need help to end refactoring branch. For now master.

**Comment by KostyaSha** [comment]
@aheritier refactoring was done based on old crowd libraries and new spring-security. 
Rememberme works fine with non SSO authentication, but i failed to find possible 1 function for validating token!

**Comment by drazzib** [comment]
Hi there !

Is there any pending issues before merging this PR ?
This leads to Jenkins using as much as 100% very quickly on our setup (many remote API calls which needs crowd2 authentication)

Regards,

**Comment by aheritier** [comment]
Hi @drazzib 
  
  Yes I few things to do on my side. I will see to raise the priority of this task. You are the second user contacting me today about this issue. Did you upgrade you jenkins core and it is getting worse than before ?
  Meantime you can find a SNAPSHOT I use in production for 2 months here : https://www.dropbox.com/s/5yk1jrgqai7g0s0/crowd2.hpi?dl=0

BR,

**Comment by drazzib** [comment]
Hi @aheritier 

Thanks for this quick feedback ! I can confirm that this issue is more frequent since our update to latest Jenkins LTS.

I've just updated using your provided HPI : many thanks !

Cheers,

**Comment by aheritier** [comment]
ok. not surprising, I noticed the problem in LTS > 1.554 https://issues.jenkins-ci.org/browse/JENKINS-16703

**Comment by goekhanm** [comment]
@aheritier using your Fix for JENKINS-27070 and JENKINS-16703. Works perfect! THX.

**Comment by recena** [comment]
@KostyaSha Is there any problem if I continue with this PR? /cc @aheritier

**Comment by piyush0101** [comment]
@aheritier Would like to use your fix. Cannot access the Dropbox link

**Comment by aheritier** [comment]
@piyush0101 The link didn't change and the binary is always available. Do you have some restriction to access to dropbox ?

**Comment by recena** [comment]
@KostyaSha ping? /cc @aheritier

**Comment by piyush0101** [comment]
We tried to use this plugin in our Jenkins setup. We also wanted to make sure that we are not degrading performace with the change. Fix for not accessing crowd with every HTTP request works fine but we found that the overall web transaction time is poor because we are spending more time in the JVM. Crowd calls were vastly reduced but the amount of time spent in JVM increased many folds.

Have a look at the performance chart. Reading graph from left to right, the first part (until 1:50PM) is with Crowd2 Plugin ver 1.6. 1:50PM - 2:20PM is with the build done using this pull request. Here we saw that crowd calls were gone but JVM performance degraded 10 fold. Later part of the chart was us trying to see if upgrading to 1.8 made a difference and it appears to be worse.

Blue: Time spent in JVM
Green: Time spent in Crowd calls
crowd2pluginprofiling

**Comment by piyush0101** [comment]
@aheritier ^^

**Comment by piyush0101** [comment]
@aheritier This is using the default plugin configuration.

**Comment by aheritier** [comment]
Thanks a lot @piyush0101 for these feedbacks. I need to investigate more because I didn't notice at all this behaviour at JVM level. The cache is a "poor" HashMap with few retention settings (number of entries to keep, ...) thus accessing to it instead of crowd shouldn't cost a lot so much time at JVM level. In my production deployment I didn't noticed this behaviour. Which version of Java are you using ? AFAIR I was on J7.

**Comment by piyush0101** [comment]
@aheritier I am using jdk1.7.0_51

**Comment by KostyaSha** [comment]
@recena it will take too much time for you (firstly my huge refactoring needs to be fixed for SSO installations) and i would like to try to work with @aheritier :)

**Comment by hypery2k** [comment]
any news on this issue?

**Comment by aheritier** [comment]
I rebased this PR on PR #11 thus:

* Upgrade to a recent plugin POM to allow to release
* Upgrade the requirement to Java 7 and thus Jenkins 1.625.3
JENKINS-27070: Upgrade to the latest version of crowd-integration-client supporting Java 7 and that integrates httpcomponents 4.x to avoid errors like this reported by httpclient 3.x which isn't supported (see : https://issues.apache.org/jira/browse/HTTPCLIENT-799):

```
Sep 16, 2014 6:24:27 PM org.apache.commons.httpclient.MultiThreadedHttpConnectionManager freeConnection
SEVERE: Host connection pool not found, hostConfig=HostConfiguration[host=XXXXX]
```

As the plugin isn't maintained anymore there is no active CI for it and you have to build this branch yourself or you can download the copy I uploaded in Jira https://issues.jenkins-ci.org/secure/attachment/40638/crowd2.hpi ( https://issues.jenkins-ci.org/browse/JENKINS-16703 ) cc @fbelzunc

**Comment by aheritier** [comment]
The patch JENKINS-16703 adds an option to activate a set of caches (off by default) which offers 2 configuration parameters when activated (thus all caches are configured the same):

* Cache size (default: 20 _entries_): A number of elements in the cache
* Cache TTL (default: 300 _seconds_): A Time To Live for entries in the cache (in seconds)

There are 5 caches for now to reduce the access to crowd

* A cache to verify if a user is member of a given group
* A cache of the user definition from his SSO token (when using crowd in SSO but this feature isn't working well)
* A cache of the user definition from his username
* A cache of the group definition from its name
* A cache of the user authorities (built from groups) 

The *Cache size* will have to be adapted to the size of the data stored in Crowd and used in Jenkins thus mainly the number of concurrent users accessing to jenkins and the number of groups used. The *Cache size* will impact the memory usage of the instance. While a small instance will be fine with 20 or 50 entries a large instance will probably require 200, 500 or 1000 entries.

The *Cache TTL* will control the freshness of the data in the cache and thus the delay to wait before seeing a change done in Crowd applied in Jenkins. Even a short cache of few minutes (2 or 5) reduces a lot the load on the crowd server and improve the performances on Jenkins side by removing duplicated calls (especially for authorisations) when Jenkins builds a page.

**Comment by pingunaut** [comment]
I'm going to merge this pull request to fix the performance and compatibility issues that currently open. Refactoring to enable SSO, rememberme and stuff will follow asap afterwards

**Pull request merged**